Manual play mode

berekuk opened the pull request:
Fixes #766.

Notes:
- cmd+enter/ctrl+enter submits the text editor (play icon is quite small and it's less convenient to press it every time); maybe I should document it with a tooltip on the play button?
- I extracted some components from `SquigglePlayground` because it was getting too long with these new state vars
- implementing the progress icon correctly was really tricky due to some React subtleties; I added some detailed comments on this, but hope this is temporary and we won't have to maintain it for too long (only until runs become async)
- `SquiggleChart` is memoized now, and it's important; without memoization it rerendered even without autoplays; so I had to memoize some other stuff too (`chartSettings`, `env`)
- @Hazelfire, sorry for messing up your #760 merge (unless that gets merged first)
- I wanted to add animation to the autoplay toggle, this PR is too useful already to delay, so that can be done later

OAGr commented on packages/components/src/components/CodeEditor.tsx:
```diff
@@ -46,6 +52,13 @@ export const CodeEditor: FC<CodeEditorProps> = ({
+        {
+          name: "submit",
+          bindKey: { mac: "Cmd-Enter", win: "Ctrl-Enter" },
+          exec: () => onSubmitRef.current?.(),
```
nice touch

OAGr commented:
Nice work!
Two points about the functionality:
1. When I press the "Run" button a second time, it gives the same exact results. I don't think it does a full new simulation.
2. The "run" button should probably be bright blue, instead of being greyed out.

OAGr commented on packages/components/src/components/SquigglePlayground.tsx:
```diff
@@ -420,10 +304,137 @@ export const SquigglePlayground: FC<PlaygroundProps> = ({
+    [vars.sampleCount, vars.xyPointLength]
+  );
+
+  const [playing, setPlaying] = useState(false); // used in manual play mode only
```
maybe rename to something like, ``isAutoRunning`` and ``setIsAutoRunning``

OAGr replied:
Also, I'd flag that we could probably use a reducer or something else a bit more sophisticated for the Squiggle Playground. All of this state is a bit much.

OAGr commented on packages/components/src/components/SquigglePlayground.tsx:
```diff
@@ -420,10 +304,137 @@ export const SquigglePlayground: FC<PlaygroundProps> = ({
+  playing: boolean;
+  stale: boolean;
+  onAutoplayChange: (value: boolean) => void;
+  play: () => void;
```
the above variable names could be improved. Use ``isX`` for booleans, and I'd suggest having more descriptive names. This function isn't very long, so longer names won't be that bad.

OAGr commented on packages/components/src/components/SquigglePlayground.tsx:
```diff
@@ -420,10 +304,137 @@ export const SquigglePlayground: FC<PlaygroundProps> = ({
+    }
+  }, [vars.autoplay, code, playing]);
+
+  const play = () => {
```
small note; but do you need the {} braces here?

berekuk replied:
> Small note; but do you need the {} braces here?

`setX` returns void, so it shouldn't matter in this case, but usually I don't skip `{}` if function's body is a single call, to make sure that it doesn't return anything. (E.g., sometimes the function gets passed to a component which expects `() => void`, and passing `() => SomethingElse` would fail type checks).

berekuk replied:
> Also, go for a more descriptive name, like, startPlaying. This isn't running once, it's just changing the state once.

This seems like an implementation detail to me? `play` is the function that should be called to trigger the run, and it effectively does that (by causing `useEffect`s to fire), and places in the code where it's used shouldn't care how the run is triggered.

I know that it doesn't cause the re-run if code hasn't changed, but that's in TODO and I hope to fix it soon.

OAGr commented on packages/components/src/components/SquigglePlayground.tsx:
```diff
@@ -477,20 +498,32 @@ export const SquigglePlayground: FC<PlaygroundProps> = ({
+            </StyledTab.List>
+            <PlayControls
+              autoplay={vars.autoplay || false}
+              stale={!vars.autoplay && renderedCode !== code}
```
It would be nice if this sort of function were part of the reducer/state object.

OAGr reviewed (approved):
I think this is a good improvement, but think that the ``play`` logic could probably be cleaned up a fair bit. I'll leave it to you to decide if you want to improve it now or in a future story.

berekuk commented:
> I think this is a good improvement, but think that the play logic could probably be cleaned up a fair bit. I'll leave it to you to decide if you want to improve it now or in a future story.

I agree, I'll do the cleanups first.

berekuk commented:
I did some renames and extracted the most complex stuff into a custom `useRunnerState` hook. React's `useReducer` won't be a good fit — we should colocate `useEffect` calls with state logic, and `useReducer` is better for pure state changes.

I also added toggle animation with framer-motion (it's +30k gzipped to our bundle, but vega weights more, and framer is really nice to have for other purposes).

Also:
- rebased from develop (had to force-push, merging was too tough)
- renamed some vars, play -> run, isX prefixes

OAGr commented:
Nice work, merging!
